packages/flutter/lib/src/widgets/app.dart

```diff
@@ -690,35 +718,135 @@ class _WidgetsAppState extends State<WidgetsApp> implements WidgetsBindingObserv
+      languageAndScriptLocales[ui.hashValues(locale.languageCode.hashCode, locale.scriptCode.hashCode)] ??= locale;
+      languageLocales[locale.languageCode.hashCode] ??= locale;
+      countryLocales[locale.countryCode.hashCode] ??= locale;
+    }
 
+    // Since languageCode-only matches are possibly low quality, we don't return
+    // it instantly when we find such a match. We check to see if the next
+    // preferred locale in the list has a high accuracy match, and only return
+    // the languageCode-only match when a higher accuracy match in the next
+    // preferred locale cannot be found.
     Locale matchesLanguageCode;
-    for (Locale locale in supportedLocales) {
-      if (locale == newLocale)
-        return newLocale;
-      if (locale.languageCode == newLocale.languageCode)
-        matchesLanguageCode ??= locale;
+    Locale matchesCountryCode;
+    // Loop over user's preferred locales
+    for (int localeIndex = 0; localeIndex < preferredLocales.length; localeIndex += 1) {
+      final Locale userLocale = preferredLocales[localeIndex];
+      // Look for perfect match.
+      if (allSupportedLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode, userLocale.countryCode.hashCode))) {
+        return userLocale;
+      }
+      // Look for language+script match.
+      if (userLocale.scriptCode != null && languageAndScriptLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode))) {
+        return languageAndScriptLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.scriptCode.hashCode)];
+      }
+      // Look for language+country match.
+      if (userLocale.countryCode != null && languageAndCountryLocales.containsKey(ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode))) {
+        return languageAndCountryLocales[ui.hashValues(userLocale.languageCode.hashCode, userLocale.countryCode.hashCode)];
+      }
+      // If there was a languageCode-only match in the previous iteration's higher
+      // ranked preferred locale, we return it if the current userLocale does not
+      // have a better match.
+      if (matchesLanguageCode != null) {
+        return matchesLanguageCode;
+      }
+      // Look and store language-only match.
+      if (languageLocales.containsKey(userLocale.languageCode.hashCode)) {
+        matchesLanguageCode = languageLocales[userLocale.languageCode.hashCode];
+        // Since first (default) locale is usually highly preferred, we will allow
+        // a languageCode-only match to be instantly matched. If the next preferred
+        // languageCode is the same, we defer hastily returning until the next iteration
+        // since at worst it is the same and at best an improved match.
+        if (localeIndex == 0
+          && !(localeIndex + 1 < preferredLocales.length
```

Comment:
I believe flutter style indention here is to align the lines and put the && at the end of the line.